aptos-move/mvhashmap/src/unsync_map.rs

```diff
@@ -1,46 +1,128 @@
     map: RefCell<HashMap<K, (Arc<V>, Option<HashValue>)>>,
+    group_cache: RefCell<HashMap<K, HashMap<T, Arc<V>>>>,
     executable_cache: RefCell<HashMap<HashValue, Arc<X>>>,
     executable_bytes: RefCell<usize>,
 }
 
-impl<K: ModulePath + Hash + Clone + Eq, V: TransactionWrite, X: Executable> Default
-    for UnsyncMap<K, V, X>
+impl<
+        K: ModulePath + Hash + Clone + Eq,
+        T: Hash + Clone + Debug + Eq + Serialize,
+        V: TransactionWrite,
+        X: Executable,
+    > Default for UnsyncMap<K, T, V, X>
 {
     fn default() -> Self {
         Self {
             map: RefCell::new(HashMap::new()),
+            group_cache: RefCell::new(HashMap::new()),
             executable_cache: RefCell::new(HashMap::new()),
             executable_bytes: RefCell::new(0),
         }
     }
 }
 
-impl<K: ModulePath + Hash + Clone + Eq, V: TransactionWrite, X: Executable> UnsyncMap<K, V, X> {
+impl<
+        K: ModulePath + Hash + Clone + Eq,
+        T: Hash + Clone + Debug + Eq + Serialize,
+        V: TransactionWrite,
+        X: Executable,
+    > UnsyncMap<K, T, V, X>
+{
     pub fn new() -> Self {
-        Self {
-            map: RefCell::new(HashMap::new()),
-            executable_cache: RefCell::new(HashMap::new()),
-            executable_bytes: RefCell::new(0),
-        }
+        Self::default()
+    }
+
+    pub fn provide_group_base_values(
+        &self,
+        group_key: K,
+        base_values: impl IntoIterator<Item = (T, V)>,
+    ) {
+        assert!(
+            self.group_cache
+                .borrow_mut()
+                .insert(
+                    group_key,
+                    base_values
+                        .into_iter()
+                        .map(|(t, v)| (t, Arc::new(v)))
+                        .collect()
+                )
+                .is_none(),
+            "UnsyncMap group cache must be empty to provide base values"
+        );
+    }
+
+    pub fn get_group_size(&self, group_key: &K) -> anyhow::Result<GroupReadResult> {
+        Ok(match self.group_cache.borrow().get(group_key) {
+            Some(group_map) => GroupReadResult::Size(group_size_as_sum(
+                group_map
+                    .iter()
+                    .flat_map(|(t, v)| v.bytes().map(|bytes| (t, bytes))),
+            )?),
+            None => GroupReadResult::Uninitialized,
+        })
+    }
+
+    pub fn get_value_from_group(&self, group_key: &K, value_tag: &T) -> GroupReadResult {
+        self.group_cache.borrow().get(group_key).map_or(
+            GroupReadResult::Uninitialized,
+            |group_map| {
+                GroupReadResult::Value(
+                    group_map
+                        .get(value_tag)
+                        .map(|v| v.extract_raw_bytes())
+                        .flatten(),
+                )
+            },
+        )
+    }
+
+    /// Contains the latest group ops (excluding deletions) for the given group key.
```

Comment:
I see "get" and "insert" methods. There doesn't seem to be any function to delete a resource in a group.